tests/test_qp.py

```diff
@@ -22,317 +24,245 @@ class QPTest(PyverbsAPITestCase):
-                        if u.is_eth(ctx, i) and u.is_root():
-                            qia.qp_type = e.IBV_QPT_RAW_PACKET
-                            try:
-                                with QP(pd, qia) as qp:
-                                    assert qp.qp_state == e.IBV_QPS_RESET, 'Raw Packet QP should have been in RESET'
-                            except PyverbsRDMAError as ex:
-                                if ex.error_code == errno.EOPNOTSUPP:
-                                    raise unittest.SkipTest("Create Raw Packet QP is not supported")
-                                raise ex
-
-    def test_create_qp_with_attr_connected(self):
-        """
-        Test QP creation via ibv_create_qp without a QPAttr object proivded.
-        Checked QP types are RC and UC.
+                        qia.qp_type = qp_type
+                        creator = pd
+
+                    qp = self.create_qp(creator, qia, is_ex, with_attr)
+                    assert qp.qp_state == qp_state , f'{get_qp_type_str(qp_type)} QP should have been in {get_qp_state_str(qp_state)}'
+
+    def test_create_rc_qp_no_attr(self):
+        """
+        Test RC QP creation via ibv_create_qp without a QPAttr object proivded.
+        """
+        self.create_qp_common_test(e.IBV_QPT_RC, e.IBV_QPS_RESET, False, False)
+
+    def test_create_uc_qp_no_attr(self):
+        """
+        Test UC QP creation via ibv_create_qp without a QPAttr object proivded.
+        """
+        self.create_qp_common_test(e.IBV_QPT_UC, e.IBV_QPS_RESET, False, False)
+
+    def test_create_ud_qp_no_attr(self):
+        """
+        Test UD QP creation via ibv_create_qp without a QPAttr object proivded.
+        """
+        self.create_qp_common_test(e.IBV_QPT_UD, e.IBV_QPS_RESET, False, False)
+
+    def test_create_raw_qp_no_attr(self):
+        """
+        Test RAW Packet QP creation via ibv_create_qp without a QPAttr object
+        proivded.
+        Raw Packet is skipped for non-root users / Infiniband link layer.
+        """
+        self.create_qp_common_test(e.IBV_QPT_RAW_PACKET, e.IBV_QPS_RESET, False, False)
+
+    def test_create_rc_qp_with_attr(self):
+        """
+        Test RC QP creation via ibv_create_qp with a QPAttr object proivded.
         """
+        self.create_qp_common_test(e.IBV_QPT_RC, e.IBV_QPS_INIT, False, True)
+
+    def test_create_uc_qp_with_attr(self):
+        """
+        Test UC QP creation via ibv_create_qp with a QPAttr object proivded.
+        """
+        self.create_qp_common_test(e.IBV_QPT_UC, e.IBV_QPS_INIT, False, True)
+
+    def test_create_ud_qp_with_attr(self):
+        """
+        Test UD QP creation via ibv_create_qp with a QPAttr object proivded.
+        """
+        self.create_qp_common_test(e.IBV_QPT_UD, e.IBV_QPS_RTS, False, True)
+
+    def test_create_raw_qp_with_attr(self):
+        """
+        Test RAW Packet QP creation via ibv_create_qp with a QPAttr object
+        proivded.
```

Comment:
type: provided*
(apply on the other docstrings)

Reply:
Fixed.